File: src/lib/sense_hat/stick.js
/**
 * Internal SenseHat Module for reading and writing values from
 * JavaScript World to the Python World. This modules set ups
 * the commmunication and allows to read and write pixels. If the
 * "Sk.sense_hat_emit" config is present, we emit events when
 * values are changed: Python -> JavaScript
 */
var $builtinmodule = function (name) {
    var mod = {};
    
    /**
     * Named InputEvent tuple
     */
    var input_event_fields = {
        "timestamp": "", 
        "key": "", 
        "state": "", 
    };
    var input_event_f = Sk.builtin.make_structseq('SenseStick', 'InputEvent', input_event_fields);
    mod.InputEvent = Sk.builtin.make_structseq('SenseStick', 'InputEvent', input_event_fields);

    var senseStickClass = function ($gbl, $loc) {
        $loc.__init__ = new Sk.builtin.func(function (self) {
            // internal value storage
            self.v = {
                _stick_file: Sk.sense_hat.sensestick,
                _eventQueue: []
            }; 

            // static class props
            // not working!
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('SENSE_HAT_EVDEV_NAME'), new Sk.builtin.str('Raspberry Pi Sense HAT Joystick'));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('EVENT_FORMAT'), new Sk.builtin.str('llHHI'));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('EVENT_SIZE'), new Sk.builtin.int_(-1)); // undefined

            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('EV_KEY'), new Sk.builtin.int_(0x01));

            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('STATE_RELEASE'), new Sk.builtin.int_(0));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('STATE_PRESS'), new Sk.builtin.int_(1));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('STATE_HOLD'), new Sk.builtin.int_(2));

            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('KEY_UP'), new Sk.builtin.int_(103));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('KEY_LEFT'), new Sk.builtin.int_(105));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('KEY_RIGHT'), new Sk.builtin.int_(106));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('KEY_DOWN'), new Sk.builtin.int_(108));
            Sk.abstr.objectSetItem(self['$d'], new Sk.builtin.str('KEY_ENTER'), new Sk.builtin.int_(28));
        })

        $loc.tp$getattr = Sk.builtin.object.prototype.GenericGetAttr;

        $loc.__setattr__ = Sk.builtin.object.prototype.GenericSetAttr;

        $loc.__str__ = new Sk.builtin.func(function (self) {
            return new Sk.builtin.str('SenseStick')
        });

        $loc.__repr__ = new Sk.builtin.func(function (self) {
            return new Sk.builtin.str('SenseStick')
        });

        $loc.close = new Sk.builtin.func(function (self) {
            // close...
            return Sk.builtin.none.none$;
        });

        $loc.__exit__ = new Sk.builtin.func(function (self) {
            return Sk.misceval.callsim(mod.SenseStick.close, self);
        });

        $loc.__enter__ = new Sk.builtin.func(function (self) {
            // nothing todo here
            return self;
        });

        $loc._stick_device = new Sk.builtin.func(function (self) {
            // special skulpt handling, we do not have a real input device!
            throw new Sk.builtin.AttributeError('Calling internal method not allowed');
        });

        /**
         * Our read impl. is a little bit smaller than the original one, as we do not support
         * other event keys then sensestick ones. So, there is not need for iterating until a
         * sensestick event has occured.
         */
        $loc.read = new Sk.builtin.func(function (self) {
            var inputEvent;
            var susp = new Sk.misceval.Suspension();
            susp.resume = function () {
                // Should the post image get stuff go here??
                if (susp.data["error"]) {
                    if (susp.data.error === 'KeyboardInterrupt') {
                        throw new Error('KeyboardInterrupt');
                    } else {
                        throw new Sk.builtin.IOError('SenseStickDevice Error');
                    }
                }

                var tup = new mod.InputEvent([
                    Sk.builtin.int_(inputEvent.timestamp),
                    Sk.builtin.int_(inputEvent.key),
                    Sk.builtin.int_(inputEvent.state)
                ]);
                return tup;
            };
            susp.data = {
                type: "Sk.promise",
                promise: new Promise(function (resolve, reject) {
                    // Read from internal eventQueue
                    if (self.v._eventQueue.length > 0) {
                        inputEvent = self.v._eventQueue.pop();
                        resolve();
                    } else {
                        // add eventlistener
                        self.v._stick_file.once('sensestick.input', function (event, inputData) {
                            // Interrupt handling, so that we do not need to wait until the users inputs something
                            if (inputData.type === 'keyboardinterrupt') {
                                reject('KeyboardInterrupt');
                            }

                            inputEvent = inputData;
                            resolve();
                        });
                    }
                })
            };
            return susp;
        });

        /**
         * Wait until our sense stick file descriptor is ready for reading (has new data!)
         * 
         * When the timeout is specified, we return early: https://github.com/RPi-Distro/python-sense-hat/blob/master/sense_hat/stick.py#L73
         */
        $loc.wait = new Sk.builtin.func(function (self, timeout) {
            var _timeout;
            if (!timeout || timeout instanceof Sk.builtin.none) {
                _timeout = null;
            } else if (Sk.builtin.checkNumber(timeout)) {
                _timeout = Sk.ffi.remapToJs(timeout);
            }

            var timeoutHandle;
            var hasEvent = false;
            var susp = new Sk.misceval.Suspension();
            susp.resume = function () {
                // Should the post image get stuff go here??
                if (susp.data["error"]) {
                    if (susp.data.error === 'KeyboardInterrupt') {
                        throw new Error('KeyboardInterrupt');
                    } else {
                        throw new Sk.builtin.IOError('SenseStickDevice Error');
                    }
                }
                return Sk.builtin.bool(hasEvent);
            };
            susp.data = {
                type: "Sk.promise",
                promise: new Promise(function (resolve, reject) {
                    //if (Sk.sense_hat.sensestick.hasEvent()) {
                    //    hasEvent = true;
                    //    resolve();
                    //} else {
                        // Listen to new one, once
                        function handleKeyInput (event, inputData) {
                            
                            // Clear timeout
                            if (timeoutHandle) {
                                window.clearTimeout(timeoutHandle);
                            }

                            if (inputData.type === 'keyboardinterrupt') {
                                reject('KeyboardInterrupt');
                            }

                            // Store event in the internal queue
                            self.v._eventQueue.push(inputData);

                            hasEvent = true; // Set return value
                            resolve();
                        }

                        self.v._stick_file.once('sensestick.input', handleKeyInput);

                        timeoutHandle = setTimeout(function() {
                            self.v._stick_file.off('sensestick.input', handleKeyInput);
                            hasEvent = false; // Timeout passed before callback occured
                            resolve()
                        }, _timeout * 1000);
                    //}
                })
            };
            return susp;
        });

        $loc.__iter__ = new Sk.builtin.func(function (self) {

            // can read?

            // 

            throw new Sk.builtin.AttributeError('Not implemented');
        });
    };

    mod.SenseStick = Sk.misceval.buildClass(mod, senseStickClass, 'SenseStick', []);

    return mod;
};

File: src/lib/sense_hat/stick.py
from _internal_sense_hat import InputEvent, _wait, _read, _start_stick_thread, _stop_stick_thread, _inspectFunction

DIRECTION_UP     = 'up'
DIRECTION_DOWN   = 'down'
DIRECTION_LEFT   = 'left'
DIRECTION_RIGHT  = 'right'
DIRECTION_MIDDLE = 'middle'

ACTION_PRESSED  = 'pressed'
ACTION_RELEASED = 'released'
ACTION_HELD     = 'held'

class SenseStick(object):
    """
    Represents the joystick on the Sense HAT.
    """
    SENSE_HAT_EVDEV_NAME = 'Raspberry Pi Sense HAT Joystick'
    EVENT_FORMAT = str('llHHI')
    #EVENT_SIZE = struct.calcsize(EVENT_FORMAT)

    EV_KEY = 0x01

    STATE_RELEASE = 0
    STATE_PRESS = 1
    STATE_HOLD = 2

    KEY_UP = 103
    KEY_LEFT = 105
    KEY_RIGHT = 106
    KEY_DOWN = 108
    KEY_ENTER = 28

    def __init__(self):
        self._callbacks = {}
        #self._stick_file = io.open(self._stick_device(), 'rb', buffering=0)
        self._callback_thread = False
        #self._callback_event = Event()

    def close(self):
        pass

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, exc_tb):
        self.close()

    def _stick_device(self):
        """
        Discovers the filename of the evdev device that represents the Sense
        HAT's joystick.
        """
        pass

    def _key_to_direction(self, key):
        return {
            self.KEY_UP:    DIRECTION_UP,
            self.KEY_DOWN:  DIRECTION_DOWN,
            self.KEY_LEFT:  DIRECTION_LEFT,
            self.KEY_RIGHT: DIRECTION_RIGHT,
            self.KEY_ENTER: DIRECTION_MIDDLE,
        }[key]

    def _state_to_action(self, state):
        return {
            self.STATE_PRESS:   ACTION_PRESSED,
            self.STATE_RELEASE: ACTION_RELEASED,
            self.STATE_HOLD:    ACTION_HELD,
        }[state]

    def _read(self):
        """
        Reads a single event from the joystick, blocking until one is
        available. Returns `None` if a non-key event was read, or an
        `InputEvent` tuple describing the event otherwise.
        """

        # Read from internal sense_hat module
        event = _read()
        (timestamp, key, state, key_type) = event
        if key_type == self.EV_KEY:
            return InputEvent([
                timestamp,
                self._key_to_direction(key),
                self._state_to_action(state)
                ])
        else:
            return None

    def _wait(self, timeout=None):
        """
        Waits *timeout* seconds until an event is available from the
        joystick. Returns `True` if an event became available, and `False`
        if the timeout expired.
        """
        # Return from internal sense_hat module
        return _wait(timeout)

    def _wrap_callback(self, fn):
        return fn
        """
        # Shamelessley nicked (with some variation) from GPIO Zero :)
        @wraps(fn)
        def wrapper(event):
            return fn()

        if fn is None:
            return None
        elif not callable(fn):
            raise ValueError('value must be None or a callable')
        elif inspect.isbuiltin(fn):
            # We can't introspect the prototype of builtins. In this case we
            # assume that the builtin has no (mandatory) parameters; this is
            # the most reasonable assumption on the basis that pre-existing
            # builtins have no knowledge of InputEvent, and the sole parameter
            # we would pass is an InputEvent
            return wrapper
        else:
            # Try binding ourselves to the argspec of the provided callable.
            # If this works, assume the function is capable of accepting no
            # parameters and that we have to wrap it to ignore the event
            # parameter
            try:
                inspect.getcallargs(fn)
                return wrapper
            except TypeError:
                try:
                    # If the above fails, try binding with a single tuple
                    # parameter. If this works, return the callback as is
                    inspect.getcallargs(fn, ())
                    return fn
                except TypeError:
                    raise ValueError(
                        'value must be a callable which accepts up to one '
                        'mandatory parameter')
        """

    def _start_stop_thread(self):
        if self._callbacks and not self._callback_thread:
            #self._callback_event.clear()
            #self._callback_thread = Thread(target=self._callback_run)
            #self._callback_thread.daemon = True
            #self._callback_thread.start()
            _start_stick_thread(self._callback_run)
            self._callback_thread = True
        elif not self._callbacks and self._callback_thread:
            #self._callback_event.set()
            _stop_stick_thread()
            self._callback_thread = False

    def _callback_run(self):
        # Use this as the internal callback function from our
        # JavaScript event based callback system
        event = self._read()
        if event:
            callback = self._callbacks.get(event.direction)
            if callback:
                argsLength = _inspectFunction(callback)
                if argsLength == 1:
                    callback(event)
                else:
                    callback()

            callback = self._callbacks.get('*')
            if callback:
                argsLength = _inspectFunction(callback)
                if argsLength == 1:
                    callback(event)
                else:
                    callback()

    def wait_for_event(self, emptybuffer=False):
        """
        Waits until a joystick event becomes available.  Returns the event, as
        an `InputEvent` tuple.

        If *emptybuffer* is `True` (it defaults to `False`), any pending
        events will be thrown away first. This is most useful if you are only
        interested in "pressed" events.
        """
        if emptybuffer:
            while self._wait(0):
                self._read()
        while self._wait():
            event = self._read()
            if event:
                return event

    def get_events(self):
        """
        Returns a list of all joystick events that have occurred since the last
        call to `get_events`. The list contains events in the order that they
        occurred. If no events have occurred in the intervening time, the
        result is an empty list.
        """
        result = []
        while self._wait(0):
            event = self._read()
            if event:
                result.append(event)
        # Make sure returned events are sorted by timestamp
        results = sorted(result, key=lambda x: x.timestamp)
        return results

    @property
    def direction_up(self):
        """
        The function to be called when the joystick is pushed up. The function
        can either take a parameter which will be the `InputEvent` tuple that
        has occurred, or the function can take no parameters at all.
        """
        return self._callbacks.get(DIRECTION_UP)


    @direction_up.setter
    def direction_up(self, value):
        self._callbacks[DIRECTION_UP] = self._wrap_callback(value)
        self._start_stop_thread()

    @property
    def direction_down(self):
        """
        The function to be called when the joystick is pushed down. The
        function can either take a parameter which will be the `InputEvent`
        tuple that has occurred, or the function can take no parameters at all.

        Assign `None` to prevent this event from being fired.
        """
        return self._callbacks.get(DIRECTION_DOWN)

    @direction_down.setter
    def direction_down(self, value):
        self._callbacks[DIRECTION_DOWN] = self._wrap_callback(value)
        self._start_stop_thread()

    @property
    def direction_left(self):
        """
        The function to be called when the joystick is pushed left. The
        function can either take a parameter which will be the `InputEvent`
        tuple that has occurred, or the function can take no parameters at all.

        Assign `None` to prevent this event from being fired.
        """
        return self._callbacks.get(DIRECTION_LEFT)

    @direction_left.setter
    def direction_left(self, value):
        self._callbacks[DIRECTION_LEFT] = self._wrap_callback(value)
        self._start_stop_thread()

    @property
    def direction_right(self):
        """
        The function to be called when the joystick is pushed right. The
        function can either take a parameter which will be the `InputEvent`
        tuple that has occurred, or the function can take no parameters at all.

        Assign `None` to prevent this event from being fired.
        """
        return self._callbacks.get(DIRECTION_RIGHT)

    @direction_right.setter
    def direction_right(self, value):
        self._callbacks[DIRECTION_RIGHT] = self._wrap_callback(value)
        self._start_stop_thread()

    @property
    def direction_middle(self):
        """
        The function to be called when the joystick middle click is pressed. The
        function can either take a parameter which will be the `InputEvent` tuple
        that has occurred, or the function can take no parameters at all.

        Assign `None` to prevent this event from being fired.
        """
        return self._callbacks.get(DIRECTION_MIDDLE)

    @direction_middle.setter
    def direction_middle(self, value):
        self._callbacks[DIRECTION_MIDDLE] = self._wrap_callback(value)
        self._start_stop_thread()

    @property
    def direction_any(self):
        """
        The function to be called when the joystick is used. The function
        can either take a parameter which will be the `InputEvent` tuple that
        has occurred, or the function can take no parameters at all.

        This event will always be called *after* events associated with a
        specific action. Assign `None` to prevent this event from being fired.
        """
        return self._callbacks.get('*')

    @direction_any.setter
    def direction_any(self, value):
        self._callbacks['*'] = self._wrap_callback(value)
        self._start_stop_thread()
